rabbitmq/rabbitmq-stream-dotnet-client
# Add AddressResolver class to handle a proxy/lb

Closes #44

What in this PR:

IClient Interface
--
Add `IClient` class. The interface for the `Client`. 

It will be useful for the unit-tests where the integration tests are hard to reproduce in certain situations. (see load balancer)
 `IClient` contains only a few methods atm, the aim is to have all the methods.

RoutingClass
--

Add RoutingClient modules with:
 - `IRouting`:  Interface for  `Routing` and tests implementations
 - `Routing` : To create the `Client`
 - `RoutingHelper<T> where T : RoutingBase` an helper to find lookup the the right node. 

Add RoutingHelper. It contains two methods:
- `LookupLeaderConnection` to create a connection to the leader node. 
- `LookupRandomConnection` Gets a random connection. The consumer can connect to a replica or leader.
 
The helper can work with or without `AddressResolver`. 

UnitTests
--
Add `UnitTests`:
 - with `LoadBalancerRouting : IRouting`
 - with `MissingFieldsRouting : IRouting `
 - with `ReplicaRouting : IRouting `

Classes to simulate the Loadbalancer situation with `AddressResolver`

How to test
---

You can create simulate a tls rabbitmq cluster with ha proxy using:
```
git clone git@github.com:rabbitmq/rabbitmq-stream-go-client.git .
make rabbitmq-ha-proxy 
```
ports:

```
 - localhost:5553 #standard stream port
 - localhost:5554 #TLS stream port
 - http://localhost:15673 #management port
```

see https://github.com/rabbitmq/rabbitmq-stream-go-client/tree/main/compose for details.

Client code:
```csharp
        private static async Task Proxy()
        {
            var addressResolver = new AddressResolver(new IPEndPoint(IPAddress.Loopback, 5554));
            var config = new StreamSystemConfig
            {
                UserName = "guest",
                Password = "guest",
                VirtualHost = "/",
                Ssl = new SslOption()
                {
                    Enabled = true,
                    AcceptablePolicyErrors = SslPolicyErrors.RemoteCertificateNotAvailable |
                                             SslPolicyErrors.RemoteCertificateChainErrors |
                                             SslPolicyErrors.RemoteCertificateNameMismatch
                },
                AddressResolver = addressResolver,
                Endpoints = new List<EndPoint> {addressResolver.EndPoint},
            };
            // Connect to the broker 

            var system = await StreamSystem.Create(config);
            var stream = Guid.NewGuid().ToString();

            await system.CreateStream(new StreamSpec(stream));
          
          
            var producer = await system.CreateProducer(
                new ProducerConfig
                {
                    Reference = Guid.NewGuid().ToString(),
                    Stream = stream,
                    // Here you can receive the messages confirmation
                    // it means the message is stored on the server
                    ConfirmHandler = conf => { Console.WriteLine($"message: {conf.PublishingId} - Code {conf.Code}"); },
                });

            // Publish the messages and set the publishingId that
            // should be sequential
            for (ulong i = 0; i < 100; i++)
            {
                var readonlySequence = new ReadOnlySequence<byte>(Encoding.UTF8.GetBytes($"hello {i}"));
                var message = new Message(new Data(readonlySequence));
                await producer.Send(i, message);
            }

            // just to show the confirmation
            Thread.Sleep(1000);

            var consumer = await system.CreateConsumer(
                new ConsumerConfig
                {
                    Reference = Guid.NewGuid().ToString(),
                    Stream = stream,
                    // Consume the stream from the beginning 
                    // See also other OffsetSpec 
                    OffsetSpec = new OffsetTypeFirst(),
                    // Receive the messages
                    MessageHandler = async (consumer, ctx, message) =>
                    {
                        Console.WriteLine(
                            $"message: {Encoding.Default.GetString(message.Data.Contents.ToArray())} - consumed");
                        await Task.CompletedTask;
                    }
                });
            Console.WriteLine($"Press to stop");
            Console.ReadLine();

            await producer.Close();
            await consumer.Close();
            await system.DeleteStream(stream);
            await system.Close();
            Console.WriteLine($"Done");
        }
```

Repeat it more than time. You should see the producer always connected to the leader and the consumer random from the leader or replicas.
